### DISCUSSION TITLE
Question about using cryoDRGN with relion 5

### DISCUSSION THREAD

**Discussion by frozenfas** [zhonge/cryodrgn#394] [Q&A] [answered]
Hello, I am new to using cryoDRGN with cryoET.

I was just wondering if cryoDRGN can be used with RELION5 and if you have a workflow.  I tried quickly to parse the run_data.star (particles) from a Refine3D job and ran into an error.  I think it is because this star file has an additional table:
`# version 50001

data_general

_rlnTomoSubTomosAre2DStacks                       1`

in addition to the statard data_optics and and data_particles table. If I remove the data_general table I can parse the poses fine.  Is this okay to do? Or am I trying to parse the wrong file?

My next question is do I have to parse the CTF information if the 2D stacks are CTF corrected in relion 5?

**Comment by frozenfas** [comment]
Regarding the error about the extra tables (i.e. data_general).  I tried to install v3.4.0 from github and this time I got the error:

`Traceback (most recent call last):
  File "/home/sconnell/local/miniconda3/envs/cryodrgn/bin/cryodrgn", line 8, in <module>
    sys.exit(main_commands())
  File "/home/sconnell/local/miniconda3/envs/cryodrgn/lib/python3.9/site-packages/cryodrgn/command_line.py", line 77, in main_commands
    _get_commands(
  File "/home/sconnell/local/miniconda3/envs/cryodrgn/lib/python3.9/site-packages/cryodrgn/command_line.py", line 72, in _get_commands
    args.func(args)
  File "/home/sconnell/local/miniconda3/envs/cryodrgn/lib/python3.9/site-packages/cryodrgn/commands/parse_pose_star.py", line 53, in main
    starfile = Starfile(args.input)
  File "/home/sconnell/local/miniconda3/envs/cryodrgn/lib/python3.9/site-packages/cryodrgn/starfile.py", line 129, in __init__
    data, data_optics = parse_star(starfile)
  File "/home/sconnell/local/miniconda3/envs/cryodrgn/lib/python3.9/site-packages/cryodrgn/starfile.py", line 22, in parse_star
    raise ValueError("Multiple data blocks detected!")
ValueError: Multiple data blocks detected!
`

**Comment by michal-g** [comment]
Hey, thanks for taking the time to test with two separate cryoDRGN versions, it helps a lot with figuring out what's going on!

The `data_general` table causes problems in older versions of cryoDRGN as we didn't fully support RELION formats beyond 3.0; thus we assume that all data is stored in a table named `data_`, `data_particles`, etc. In the pre-releases of v3.4.0 we are in the middle of adding support for these newer RELION formats, and in particular those with a `data_optics` table; however we had not yet considered files with other tables like `data_general`, hence the second error above.

Your intuition that we can ignore the `data_general` table is correct; in the latest v3.4.0 pre-release I've updated the code for parsing .star input to better handle these cases, which you can install using:
```pip install -i https://test.pypi.org/simple/ --extra-index-url https://pypi.org/simple/ 'cryodrgn<=3.4.0' --pre```

Please let us know if you find any other problems, and thanks again for participating in the beta testing for v3.4.0!

**Reply by frozenfas** [reply]
Thanks.  I reinstalled the v3.4.0 pre-release as you suggested, and it does parse the pose data from run_data.star without error.

If I try to move to the next step:
`cryodrgn train_vae Refine3D/job150/run_data_modcryodrgn.star  --poses cryodrgn/pose.pkl  --encode-mode tilt --dose-per-tilt 1.8 --zdim 8 -n 50 --beta 0.025 -o cryodrgn/`

I get the following error:
`(INFO) (train_vae.py) (27-Aug-24 14:25:07) /home/sconnell/local/miniconda3/envs/cryodrgn/bin/cryodrgn train_vae Refine3D/job150/run_data_modcryodrgn.star --poses cryodrgn/pose.pkl --encode-mode tilt --dose-per-tilt 1.8 --zdim 8 -n 50 --beta 0.025 -o cryodrgn/
(INFO) (train_vae.py) (27-Aug-24 14:25:07) cryoDRGN 3.4.0b0
(INFO) (train_vae.py) (27-Aug-24 14:25:07) Namespace(particles='/mnt/McQueen-002/sconnell/EMDB/EMPIAR-10860-Ecoli-GOSLAR/relion/Refine3D/job150/run_data_modcryodrgn.star', outdir='/mnt/McQueen-002/sconnell/EMDB/EMPIAR-10860-Ecoli-GOSLAR/relion/cryodrgn', zdim=8, poses='/mnt/McQueen-002/sconnell/EMDB/EMPIAR-10860-Ecoli-GOSLAR/relion/cryodrgn/pose.pkl', ctf=None, load=None, checkpoint=1, log_interval=1000, verbose=False, seed=95994, ind=None, invert_data=True, window=True, window_r=0.85, datadir=None, lazy=False, shuffler_size=0, num_workers=0, max_threads=16, ntilts=10, random_tilts=False, t_emb_dim=64, tlayers=3, tdim=1024, dose_per_tilt=1.8, angle_per_tilt=3, num_epochs=50, batch_size=8, wd=0, lr=0.0001, beta='0.025', beta_control=None, norm=None, amp=True, multigpu=False, do_pose_sgd=False, pretrain=1, emb_type='quat', pose_lr=0.0003, qlayers=3, qdim=1024, encode_mode='tilt', enc_mask=None, use_real=False, players=3, pdim=1024, pe_type='gaussian', feat_sigma=0.5, pe_dim=None, domain='fourier', activation='relu', func=<function main at 0x7750c1022dc0>)
(INFO) (train_vae.py) (27-Aug-24 14:25:07) Use cuda True
(INFO) (train_vae.py) (27-Aug-24 14:25:07) Loading dataset from /mnt/McQueen-002/sconnell/EMDB/EMPIAR-10860-Ecoli-GOSLAR/relion/Refine3D/job150/run_data_modcryodrgn.star
Traceback (most recent call last):
  File "/home/sconnell/local/miniconda3/envs/cryodrgn/bin/cryodrgn", line 8, in <module>
    sys.exit(main_commands())
  File "/home/sconnell/local/miniconda3/envs/cryodrgn/lib/python3.9/site-packages/cryodrgn/command_line.py", line 80, in main_commands
    _get_commands(
  File "/home/sconnell/local/miniconda3/envs/cryodrgn/lib/python3.9/site-packages/cryodrgn/command_line.py", line 75, in _get_commands
    args.func(args)
  File "/home/sconnell/local/miniconda3/envs/cryodrgn/lib/python3.9/site-packages/cryodrgn/commands/train_vae.py", line 698, in main
    data = dataset.TiltSeriesData(  # FIXME: maybe combine with above?
  File "/home/sconnell/local/miniconda3/envs/cryodrgn/lib/python3.9/site-packages/cryodrgn/dataset.py", line 150, in __init__
    super().__init__(tiltstar, ind=ind, **kwargs)
  File "/home/sconnell/local/miniconda3/envs/cryodrgn/lib/python3.9/site-packages/cryodrgn/dataset.py", line 49, in __init__
    self.src = ImageSource.from_file(
  File "/home/sconnell/local/miniconda3/envs/cryodrgn/lib/python3.9/site-packages/cryodrgn/source.py", line 141, in from_file
    return StarfileSource(
  File "/home/sconnell/local/miniconda3/envs/cryodrgn/lib/python3.9/site-packages/cryodrgn/source.py", line 641, in __init__
    sdata[["__mrc_index", "__mrc_filename"]] = sdata["_rlnImageName"].str.split(
  File "/home/sconnell/local/miniconda3/envs/cryodrgn/lib/python3.9/site-packages/pandas/core/frame.py", line 3968, in __setitem__
    self._setitem_array(key, value)
  File "/home/sconnell/local/miniconda3/envs/cryodrgn/lib/python3.9/site-packages/pandas/core/frame.py", line 4010, in _setitem_array
    check_key_length(self.columns, key, value)
  File "/home/sconnell/local/miniconda3/envs/cryodrgn/lib/python3.9/site-packages/pandas/core/indexers/utils.py", line 401, in check_key_length
    raise ValueError("Columns must be same length as key")
ValueError: Columns must be same length as key`


Setting the --datadir in the cryodrgn train_vae does not help.  Note I did not downsample the data as I am using 96x96 2D projections.

I think the issue is that in the run_data.star my "particles" have _rlnImageName names like:
`Extract/job222/Subtomograms/ts-69/12_stack2d.mrcs`
Namely each tilt series has its own directory and each particle its own stack. I think in source.py it tries to get the index and filename from splitting the _rlnImageName using the @ symbol. 

Am I extracting the 2D particle tilt series images wrong in Relion5 to get these different file names? If it helps, I can share the required files. Is it better to pass the output of Relion5 through M to get it in a better format? 

If I am not making a mistake somewhere , maybe it is better to have a "pre-processing" script to bring the Relion5 files more in line with what cryoDRGN is expecting.

**Reply by michal-g** [reply]
> If I am not making a mistake somewhere , maybe it is better to have a "pre-processing" script to bring the Relion5 files more in line with what cryoDRGN is expecting.

I think we will aim to have a solution built into cryoDRGN reconstruction methods to avoid the need for a separate pre-processing step for the user — as per my comment below, I hope to have more updates soon!

**Comment by PeterVanF** [comment]
Hi everyone,
I encountered an issue with pre-processing the RELION 5 subtomogram file, particularly in converting the CTF information to a ctf.pkl file. Does cryoDRGN currently support this? Additionally, I noticed the following statement: "RELION 5 subtilts are CTF-premultiplied and have not been thoroughly tested in cryoDRGN-ET." Has anyone tested the performance of RELION 5 in cryoDRGN?

**Reply by michal-g** [reply]
Hey Peter, can you share the error message you saw, and also which version of cryoDRGN you were using? But yes, preliminary testing with RELION 5 has identified a number of issues with the new format!

**Comment by michal-g** [comment]
Hey all, we have now found a number of incompatibilities with the recent changes introduced in RELION 5, including replicating those that have been pointed out above, and are looking to update cryoDRGN code to resolve this in a patch release within the next few weeks.

I am also escalating this to an issue (#401) for reasons I believe are self-evident. I will have an update there shortly, and if you find any more potential problems with the RELION 5 format please let us know in the issues thread, or feel free to continue the discussion here!

-Mike

**Comment by shahpnmlab** [comment]
Hi @michal-g 
I observed some other bugs (#420). I was able to circumvent this by forcing WarpTools to export particleseries data with the `--max_missing_tilts 0`. However, it still generated `.mrcs` files for all coordinates in the input star file but generated a star file with only those particles whose views were present in every image of the tilt series. So that might be something you need to take care of when patching things at your end.

After I did all of that - i ran into the problem stated in this thread.
Best,
Pranav

**Comment by ryanfeathers** [comment]
> Hello, I am new to using cryoDRGN with cryoET.
> 
> I was just wondering if cryoDRGN can be used with RELION5 and if you have a workflow. I tried quickly to parse the run_data.star (particles) from a Refine3D job and ran into an error. I think it is because this star file has an additional table: `# version 50001
> 
> data_general
> 
> _rlnTomoSubTomosAre2DStacks 1`
> 
> in addition to the statard data_optics and and data_particles table. If I remove the data_general table I can parse the poses fine. Is this okay to do? Or am I trying to parse the wrong file?
> 
> My next question is do I have to parse the CTF information if the 2D stacks are CTF corrected in relion 5?

Hi @frozenfas 

I realize this is a little late but let me know if you are still interested in using cryoDRGN-ET with data from RELION5. I have been working on a conversion and could use some help with testing. 

Best,

Ryan

**Reply by frozenfas** [reply]
HI @ryanfeathers 

Sorry for the delay.  Yes, I would be very excited to test it out on our project.

**Reply by ryanfeathers** [reply]
Emailed you!

**Comment by ryanfeathers** [comment, marked as answer]
Support for RELION5 `--tomo --no_ctf` particles is now available in cryoDRGN 3.5.x
